File: windows/src/SoliBee.Core/Models/HoneycombRuleLocalization.cs
using System.Collections.Generic;
using SoliBee.Core.Localization;

namespace SoliBee.Core.Models;

// Display-only translations for HoneycombRule.DisplayName() / .GetExplanation() and
// HoneycombDifficulty.DisplayName() — all stay English at the source (DisplayName()
// is the persisted/business-logic source of truth, see HoneycombTypes.cs's doc
// comment; ViewModel-constructed banner text keeps using DisplayName() directly,
// untouched). Lives in Core (not Desktop) because HoneycombViewModel (Core) needs it
// too, not just the View layer. Mirrors the Mac port's HoneycombRuleLocalization.swift
// (shared/Honeycomb/Models/).
public static class HoneycombRuleLocalization
{
    public static string LocalizedRuleName(HoneycombRule rule, AppLanguage language) => rule switch
    {
        HoneycombRule.Ascension   => Strings.Get(StringKey.RuleNamePollination, language),
        HoneycombRule.Descension  => Strings.Get(StringKey.RuleNameSmokedOut, language),
        HoneycombRule.Same        => Strings.Get(StringKey.RuleNameSymmetry, language),
        HoneycombRule.Plus        => Strings.Get(StringKey.RuleNameMathBee, language),
        HoneycombRule.FallenAce   => Strings.Get(StringKey.RuleNameQueensFall, language),
        HoneycombRule.Reverse     => Strings.Get(StringKey.RuleNameInversion, language),
        HoneycombRule.AllOpen     => Strings.Get(StringKey.RuleNameClearSkies, language),
        HoneycombRule.ThreeOpen   => Strings.Get(StringKey.RuleNameScoutingParty, language),
        HoneycombRule.Swap        => Strings.Get(StringKey.RuleNameNectarExchange, language),
        HoneycombRule.Order       => Strings.Get(StringKey.RuleNameHierarchy, language),
        HoneycombRule.Chaos       => Strings.Get(StringKey.RuleNameFrenzy, language),
        HoneycombRule.BombShelter => Strings.Get(StringKey.RuleNameCappedBrood, language),
        HoneycombRule.SuddenDeath => Strings.Get(StringKey.RuleNameSwarmToTheDeath, language),
        _ => rule.DisplayName(),
    };

    // The dynamic (activeSuits non-empty) Ascension/Descension wording isn't covered
    // by a translation yet, so it falls through to GetExplanation() as-is.
    public static string LocalizedRuleExplanation(HoneycombRule rule, HashSet<string>? activeSuits, AppLanguage language) => rule switch
    {
        HoneycombRule.Same        => Strings.Get(StringKey.RuleExplanationSame, language),
        HoneycombRule.Plus        => Strings.Get(StringKey.RuleExplanationPlus, language),
        HoneycombRule.FallenAce   => Strings.Get(StringKey.RuleExplanationFallenAce, language),
        HoneycombRule.Reverse     => Strings.Get(StringKey.RuleExplanationReverse, language),
        HoneycombRule.Order       => Strings.Get(StringKey.RuleExplanationOrder, language),
        HoneycombRule.Chaos       => Strings.Get(StringKey.RuleExplanationChaos, language),
        HoneycombRule.AllOpen     => Strings.Get(StringKey.RuleExplanationAllOpen, language),
        HoneycombRule.ThreeOpen   => Strings.Get(StringKey.RuleExplanationThreeOpen, language),
        HoneycombRule.BombShelter => Strings.Get(StringKey.RuleExplanationBombShelter, language),
        HoneycombRule.SuddenDeath => Strings.Get(StringKey.RuleExplanationSuddenDeath, language),
        HoneycombRule.Swap        => Strings.Get(StringKey.RuleExplanationSwap, language),
        HoneycombRule.Ascension when activeSuits == null || activeSuits.Count == 0 => Strings.Get(StringKey.RuleExplanationAscensionGeneric, language),
        HoneycombRule.Descension when activeSuits == null || activeSuits.Count == 0 => Strings.Get(StringKey.RuleExplanationDescensionGeneric, language),
        _ => rule.GetExplanation(activeSuits),
    };

    public static string LocalizedDifficultyName(HoneycombDifficulty difficulty, AppLanguage language) => difficulty switch
    {
        HoneycombDifficulty.Easy      => Strings.Get(StringKey.StatBabyBee, language),
        HoneycombDifficulty.Medium    => Strings.Get(StringKey.StatHoneyBee, language),
        HoneycombDifficulty.Hard      => Strings.Get(StringKey.StatQueenBee, language),
        HoneycombDifficulty.UltraHard => Strings.Get(StringKey.StatKillerBee, language),
        _ => difficulty.DisplayName(),
    };

    // The player's own hand-area label, opposite the opponent's difficulty-tier name
    // (LocalizedDifficultyName above) — shows a rank tied to the player's lifetime
    // card-collection progress (HoneycombProfileManager.Shared.UnlockedCardIds)
    // instead of a static "Player" for everyone. totalCards should be
    // HoneycombDatabase.Shared.AllCards.Count, read live rather than hardcoded, so
    // this never drifts out of sync if the generated collection's size ever changes.
    // cardsCollected >= totalCards is checked before the 500+ band specifically so a
    // player who has genuinely unlocked every card gets the distinct completionist
    // title (Apiarist) rather than being folded into the same "500+" band as someone
    // who's merely close (Hive Monarch). Mirrors the Mac port's identical
    // honeycombLocalizedPlayerRankName in HoneycombRuleLocalization.swift.
    public static string LocalizedPlayerRankName(int cardsCollected, int totalCards, AppLanguage language)
    {
        if (totalCards > 0 && cardsCollected >= totalCards) return Strings.Get(StringKey.RankApiarist, language);
        if (cardsCollected >= 500) return Strings.Get(StringKey.RankHiveMonarch, language);
        if (cardsCollected >= 400) return Strings.Get(StringKey.RankSwarmLeader, language);
        if (cardsCollected >= 300) return Strings.Get(StringKey.RankRoyalAttendant, language);
        if (cardsCollected >= 200) return Strings.Get(StringKey.RankCombArchitect, language);
        if (cardsCollected >= 150) return Strings.Get(StringKey.RankWaggleDancer, language);
        if (cardsCollected >= 100) return Strings.Get(StringKey.RankGuardBee, language);
        if (cardsCollected >= 50)  return Strings.Get(StringKey.RankWorkerBee, language);
        if (cardsCollected >= 20)  return Strings.Get(StringKey.RankScoutBee, language);
        if (cardsCollected >= 10)  return Strings.Get(StringKey.RankNurseBee, language);
        return Strings.Get(StringKey.PlayerLabel, language);
    }
}

File: shared/Honeycomb/Models/HoneycombRuleLocalization.swift
import Foundation

// Display-only translations for HoneycombRule.rawValue / .explanation() and
// HoneycombDifficulty.displayName — all three stay English at the source (rawValue
// is persisted/compared elsewhere — bannedRules is a Set<String> of exact rawValues,
// and HoneycombViewModel interpolates rawValue/displayName directly into banner
// text), so those stay untouched. Takes `language` directly (not an AppCoordinator)
// so it's callable from both View code (pass coordinator.language) and ViewModel
// code, which doesn't hold an AppCoordinator reference (same reasoning as
// BannerCatalog.currentLanguage). Factored out here rather than duplicated per call
// site so they can't drift out of sync as rules are added or re-translated.
public func honeycombLocalizedRuleName(_ ruleName: String, language: AppLanguage) -> String {
    switch ruleName {
    case "Normal Mode":                       return L(.normalModeBanItem, language: language)
    case HoneycombRule.ascension.rawValue:    return L(.ruleNamePollination, language: language)
    case HoneycombRule.descension.rawValue:   return L(.ruleNameSmokedOut, language: language)
    case HoneycombRule.same.rawValue:         return L(.ruleNameSymmetry, language: language)
    case HoneycombRule.plus.rawValue:         return L(.ruleNameMathBee, language: language)
    case HoneycombRule.fallenAce.rawValue:    return L(.ruleNameQueensFall, language: language)
    case HoneycombRule.reverse.rawValue:      return L(.ruleNameInversion, language: language)
    case HoneycombRule.allOpen.rawValue:      return L(.ruleNameClearSkies, language: language)
    case HoneycombRule.threeOpen.rawValue:    return L(.ruleNameScoutingParty, language: language)
    case HoneycombRule.swap.rawValue:         return L(.ruleNameNectarExchange, language: language)
    case HoneycombRule.order.rawValue:        return L(.ruleNameHierarchy, language: language)
    case HoneycombRule.chaos.rawValue:        return L(.ruleNameFrenzy, language: language)
    case HoneycombRule.bombShelter.rawValue:  return L(.ruleNameCappedBrood, language: language)
    case HoneycombRule.suddenDeath.rawValue:  return L(.ruleNameSwarmToTheDeath, language: language)
    default: return ruleName
    }
}

// The dynamic (activeSuits non-empty) Ascension/Descension wording — with a specific
// suit list interpolated into the English sentence — isn't covered by a translation
// yet, so those two cases fall through to explanation(activeSuits:) as-is whenever
// suits are actually active; the generic (no-suits-yet) wording is translated.
public func honeycombLocalizedRuleExplanation(
    _ rule: HoneycombRule, activeSuits: Set<String> = [], language: AppLanguage
) -> String {
    switch rule {
    case .same:        return L(.ruleExplanationSame, language: language)
    case .plus:        return L(.ruleExplanationPlus, language: language)
    case .fallenAce:   return L(.ruleExplanationFallenAce, language: language)
    case .reverse:     return L(.ruleExplanationReverse, language: language)
    case .order:       return L(.ruleExplanationOrder, language: language)
    case .chaos:       return L(.ruleExplanationChaos, language: language)
    case .allOpen:     return L(.ruleExplanationAllOpen, language: language)
    case .threeOpen:   return L(.ruleExplanationThreeOpen, language: language)
    case .bombShelter: return L(.ruleExplanationBombShelter, language: language)
    case .suddenDeath: return L(.ruleExplanationSuddenDeath, language: language)
    case .swap:        return L(.ruleExplanationSwap, language: language)
    case .ascension:
        guard activeSuits.isEmpty else { return rule.explanation(activeSuits: activeSuits) }
        return L(.ruleExplanationAscensionGeneric, language: language)
    case .descension:
        guard activeSuits.isEmpty else { return rule.explanation(activeSuits: activeSuits) }
        return L(.ruleExplanationDescensionGeneric, language: language)
    }
}

public func honeycombLocalizedDifficultyName(_ difficulty: HoneycombDifficulty, language: AppLanguage) -> String {
    switch difficulty {
    case .easy:      return L(.statBabyBee, language: language)
    case .medium:    return L(.statHoneyBee, language: language)
    case .hard:      return L(.statQueenBee, language: language)
    case .ultraHard: return L(.statKillerBee, language: language)
    }
}

// The player's own hand-area label, opposite the opponent's difficulty-tier name
// (honeycombLocalizedDifficultyName above) — shows a rank tied to the player's
// lifetime card-collection progress (HoneycombProfileManager.shared.unlockedCardIds)
// instead of a static "Player" for everyone. totalCards should be
// HoneycombDatabase.shared.allCards.count, read live rather than hardcoded, so this
// never drifts out of sync if the generated collection's size ever changes.
// cardsCollected >= totalCards is checked before the 500+ band specifically so a
// player who has genuinely unlocked every card gets the distinct completionist title
// (Apiarist) rather than being folded into the same "500+" band as someone who's
// merely close (Hive Monarch).
public func honeycombLocalizedPlayerRankName(cardsCollected: Int, totalCards: Int, language: AppLanguage) -> String {
    if totalCards > 0 && cardsCollected >= totalCards {
        return L(.rankApiarist, language: language)
    }
    switch cardsCollected {
    case 500...:    return L(.rankHiveMonarch, language: language)
    case 400..<500: return L(.rankSwarmLeader, language: language)
    case 300..<400: return L(.rankRoyalAttendant, language: language)
    case 200..<300: return L(.rankCombArchitect, language: language)
    case 150..<200: return L(.rankWaggleDancer, language: language)
    case 100..<150: return L(.rankGuardBee, language: language)
    case 50..<100:  return L(.rankWorkerBee, language: language)
    case 20..<50:   return L(.rankScoutBee, language: language)
    case 10..<20:   return L(.rankNurseBee, language: language)
    default:        return L(.playerLabel, language: language)
    }
}
